File: iotServerLib/iotNode.py
#!/usr/bin/python3
# File name   : iotNode.py
# Description : base class for all IOT nodes 

class IotNode(object):
    """ the base class for all IOT nodes """
    def __init__(self, name, parent):
        """ construct an IotNode
        name - the name of the node
        parent - parent IotNode object. None for root node.
        """
        self.name = name
        self.parent = parent

    def fullPathName(self):
        """ get the node's full path name with parent path included """
        if self.parent == None:
            return self.name
        else:
            return self.parent.fullPathName() + '.' + self.name

    def root(self):
        """ get the root IotNode """
        if self.parent == None:
            return self
        else:
            return self.parent.root()

File: iotClientLib/iot/IotNode.cs
using System.Collections.Generic;

namespace IotClientLib
{
    /// <summary>
    /// Node defines the base interface for any components in IOT
    /// </summary>
    public interface IotNode
    {
        /// <summary>
        /// the ID for the node
        /// </summary>
        string Id { get; set; }

        /// <summary>
        /// the name for the node
        /// </summary>
        string Name { get; set; }

        /// <summary>
        /// The parent of this node
        /// </summary>
        IotNode Parent { get; set; }

        /// <summary>
        /// The children nodes of this node
        /// </summary>
        IList<IotNode> Children { get; }
    }
}
